### DISCUSSION TITLE
2D quality in Godot 4

### DISCUSSION THREAD

**Discussion by hydroper** [godotengine/godot-proposals#7099] [2D]
As I understand it, Godot 4 doesn't properly support anti-antialising yet. Is that going to be supported soon?

Another thing: why are fonts pixelated? Are they converted to bitmap? Is there an option to render fonts from their files (e.g. TTF) instead of converting them to bitmap? Any proposal on this? Or is that an issue only when you resize the window, for example? I guess it can be an issue since export platforms can have varying resolution (e.g. a handheld tablet vs a phone).

**Comment by AThousandShips** [comment]
> As I understand it, Godot 4 doesn't properly support anti-antialising yet. Is that going to be supported soon?

What do you mean? And where are you getting this?

> why are fonts pixelated?

Are they? Are you using bitmap fonts that are pixelated? Otherwise they are rendered via shapes and shouldn't be pixelated

Can you provide examples of them being pixelated

**Reply by hydroper** [reply]
I'm using TTFs indeed, not bitmap fonts. Here:

![image](https://github.com/godotengine/godot-proposals/assets/110610266/90f9452c-2a92-4812-b6dc-91ed228ea57d)

![image](https://github.com/godotengine/godot-proposals/assets/110610266/0086c59e-9564-4cad-910f-3b40437747cd)

![image](https://github.com/godotengine/godot-proposals/assets/110610266/ad3a5c12-da73-4051-a13d-ecfb102610e8)

![image](https://github.com/godotengine/godot-proposals/assets/110610266/e4d0370f-5410-43c1-b4c9-253e724c9749)

I tried selecting one of the anti-aliasing options, but the renderer says it's not supported yet.

**Reply by AThousandShips** [reply]
Antialiasing is supported, are you on the compatability renderer? Then it might very well not be supported, I can use anti-aliasing just fine when importing, also have you tried other means like MSDF?

**Reply by hydroper** [reply]
Right, it seems like antialiasing is supported in `forward_plus` renderer since I get no warning anymore. But the TTFs are still pixelated...

**Reply by AThousandShips** [reply]
You need to make them the correct size for displaying I'd say, just being able to scale them up and down isn't gonna be guaranteed when you use them in a bitmap format, so using other methods like MSDF would help in *those* cases, anti-aliasing only applies to when the bitmaps are created, not making them able to be scaled freely, that's not possible

**Reply by hydroper** [reply]
But TTFs aren't bitmap, they're like SVGs

**Reply by AThousandShips** [reply]
Yes, but they *are* rasterized by default AFIK, *unless* you use MSDF:
https://docs.godotengine.org/en/stable/tutorials/ui/gui_using_fonts.html

Fonts are rasterized so that they are looking appropriate for the scale they are at, you need to adapt the font size to match this I'd say

**Reply by hydroper** [reply]
I see, useful! Strange that I enabled MSDF at "Import", re-imported the font and saw no effect. The UI control I was looking at uses a theme that overrides the font for `Label` and also `RichTextLabel`. It points to a Roboto regular ttf.

**Reply by Calinou** [reply]
Note that MSDF is not required to get fonts to look smooth. The FreeType rasterizer already antialiases fonts by default, which is what DynamicFont uses.

If your visuals look pixelated, maybe you're downscaling nodes a lot or are using the `viewport` stretch mode with the window size below the default (or above the default, but with a non-integer scale factor).

**Reply by hydroper** [reply]
The UI component scales its Y from 0 to 1 when it popups and from 1 to 0 when collapsing, so no problem with node scale here. Now, the window resolution, I'm not very sure which is its value, but I suppose it should be an integer?

**Reply by Calinou** [reply]
> Now, the window resolution, I'm not very sure which is its value, but I suppose it should be an integer?

Which stretch mode are you using? If you're not using the `viewport` stretch mode, you can use any size you want. However, if using the `viewport` stretch mode, you need to make sure the window size is an integer multiple of the base size (the aspect ratio can differ if using the `keep`, `keep_width`, `keep_height` or `expand` stretch aspects). For instance, if your project's base resolution is 640×480, you should only use a window size that is 480 pixels tall, 960 pixels tall, 1440 pixels tall, etc.

See https://github.com/godotengine/godot-proposals/issues/1666 and https://tanalin.com/en/articles/integer-scaling/ for detailed rationale.

**Reply by hydroper** [reply]
I didn't know what the stretch mode meant, and I wanted `aspect = keep`, but changing from `strech_mode = viewport` to `stretch_mode = canvas_items` solved it. I think that `viewport` behavior you've described would be complicated as the window resolution can be anything.

**Comment by hydroper** [comment]
Enabling MSDF as mentioned by @AThousandShips at each TTF's _Import_  and changing the window `stretch_mode = viewport` to `stretch_mode = canvas_items`, based on @Calinou explanation, allowed the fonts to scale their shape to any size. I didn't test anti-aliasing on bitmaps yet, but I think the quality issue I had was solved.
